### PR TITLE
feat(nuxt): add `onBeforeRouteLeave` and `onBeforeRouteUpdate` composables

### LINKED ISSUE #8882: Auto import vue router navigation guards
### Describe the feature

It would be great if `onBeforeRouteLeave` and `onBeforeRouteUpdate` were available without manually importing them from `vue-router`.

### Additional information

- [ ] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://v3.nuxtjs.org/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/framework/discussions) and [issues](https://github.com/nuxt/framework/issues).

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves #8882

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This adds `onBeforeRouteLeave` and `onBeforeRouteUpdate` composables, both for universal and `vue-router` implementations.

It also hotfixes an issue where changes to preset sources were persisting across nuxt restarts. (We should likely resolve this later in kit module options resolution.)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/pages/module.ts [hunk 1]
```diff
@@ -54,6 +54,14 @@ export default defineNuxtModule({
       references.push({ types: 'vue-router' })
     })
 
+    // Add vue-router route guard imports
+    nuxt.hook('imports:sources', (sources) => {
+      const routerImports = sources.find(s => s.from === '#app' && s.imports.includes('onBeforeRouteLeave'))
```
Comment (pi0) [inline packages/nuxt/src/pages/module.ts:59]: Shouldn't `onBeforeRouteUpdate` be rewritten to `vue-router` instead?

Comment (danielroe) [reply packages/nuxt/src/pages/module.ts:59]: it does. Both are included in the same preset.

Comment (pi0) [reply packages/nuxt/src/pages/module.ts:59]: Aha! It is to find the preset 👍🏼

**Review by pi0 — APPROVED** [review]

**Comment by pi0** [comment]
Would be nice to add docs later :)

**Pull request merged**